### PR TITLE
Issue 50354: LKSM: Include indicator of unavailable file in apps

### PR DESCRIPTION
**Pull request by XingY**
#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1500
- https://github.com/LabKey/platform/pull/5526
- https://github.com/LabKey/limsModules/pull/298
- https://github.com/LabKey/testAutomation/pull/1946

#### Changes
- Append " (unavailable)" suffix for file displayValues for missing files

### CODE REVIEW

File: api/src/org/labkey/api/data/AbstractFileDisplayColumn.java [hunk 1]
```diff
@@ -77,9 +79,10 @@ protected boolean isImage(String filename)
 
     protected void renderIconAndFilename(RenderContext ctx, Writer out, String filename, @Nullable String fileIconUrl, @Nullable String popupIconUrl, boolean link, boolean thumbnail) throws IOException
     {
-        if (null != filename)
+        if (null != filename && !StringUtils.isEmpty(filename))
         {
-            String url = renderURL(ctx);
+            StringExpression s = compileExpression(ctx.getViewContext());
+            String url = null == s ? null : s.eval(ctx);
```
Comment (cnathe) [inline api/src/org/labkey/api/data/AbstractFileDisplayColumn.java]: I was confused by this change at first. However, looking at IntelliJ, it looks like you are effectively skipping over the renderURL override in DataColumn.java and reverting back to the one in DisplayColumn.java. Is that correct? If so, might be worth a comment here.

File: api/src/org/labkey/api/study/assay/FileLinkDisplayColumn.java [hunk 2]
```diff
@@ -255,6 +255,9 @@ protected String getFileName(RenderContext ctx, Object value)
             {
                 result = f.getName();
             }
+
+            if (!f.exists())
+                result += " (unavailable)";
```
Comment (cnathe) [inline api/src/org/labkey/api/study/assay/FileLinkDisplayColumn.java:265]: it looks like I am getting double "(unavailable)" messages in the LKS side now with this change. This class already adds the suffix at line 328. is that not enough to detect on the client side for the app?
Screenshot 2024-05-23 at 1 58 16 PM

**Review by cnathe — CHANGES_REQUESTED** [review]

**Review by cnathe — APPROVED** [review]

File: api/src/org/labkey/api/data/AbstractFileDisplayColumn.java (hunk 1 above)
Comment (XingY) [reply api/src/org/labkey/api/data/AbstractFileDisplayColumn.java]: Yep, comment added

File: api/src/org/labkey/api/study/assay/FileLinkDisplayColumn.java (hunk 2 above)
Comment (XingY) [reply api/src/org/labkey/api/study/assay/FileLinkDisplayColumn.java:265]: good catch. should be fixed now. The LKS side uses a different code path for display column to add the suffix.The client side needs getQuery api displayValues changed.

**Pull request merged**